src/libAtomVM/memory.h

```diff
@@ -77,22 +74,33 @@ enum MemoryGCResult memory_gc(Context *ctx, int new_size);
 term memory_copy_term_tree(term **new_heap, term t, term *mso_list);
 
 /**
- * @brief meakes sure that the given context has given free memory
+ * @brief makes sure that the given context has given free memory.
  *
- * @details this function makes sure that at least size terms are available, when not available gc will be performed, any existing term might be invalid after this call.
-
+ * @details this function makes sure at least size terms are available. Optionally,
+ * it can shrink the heap to the specified size, depending on allocation strategy.
+ * The function can also be passed roots to update during any garbage collection.
  * @param ctx the target context.
  * @param size needed available memory.
+ * @param shrink if the heap can be shrunk
+ * @param num_roots number of roots
+ * @param roots roots to preserve
  */
-enum MemoryGCResult memory_ensure_free(Context *ctx, uint32_t size) MUST_CHECK;
+enum MemoryGCResult memory_ensure_free_opt(Context *ctx, uint32_t size, enum MemoryShrinkMode shrink_mode, int num_roots, term *roots) MUST_CHECK;
 
 /**
- * @brief runs a garbage collection and shrinks used memory
+ * @brief makes sure that the given context has given free memory
  *
- * @details runs a garbage collection and shrinks used memory, a new heap will be allocated, any existing term might be invalid after this call.
- * @param ctx the context on which the garbage collection will be performed.
+ * @details this function makes sure that at least size terms are available, when not available gc will be performed, any existing term might be invalid after this call.
+ * It does not shrink the heap, so if this function is called with a given value N and it is later called with a smaller value n, the actual amount of available free
+ * memory is N.
+ * @param ctx the target context.
+ * @param size needed available memory.
  */
-enum MemoryGCResult memory_gc_and_shrink(Context *ctx) MUST_CHECK;
+static inline enum MemoryGCResult memory_ensure_free(Context *ctx, uint32_t size)
+{
+    return memory_ensure_free_opt(ctx, size, MEMORY_NO_SHRINK, 0, NULL);
+}
+MUST_CHECK
```

Comment:
MUST_CHECK at the end of the inline function is harder to notice rather than the opposite.

So I would change it to: `MUST_CHECK static inline enum MemoryGCResult memory_ensure_free(Context *ctx, uint32_t size)`.